src/shutdown/umount.c

```diff
@@ -245,6 +245,10 @@ static int loopback_list_get(MountPoint **head) {
         if (r < 0)
                 return r;
 
+        r = sd_device_enumerator_add_match_sysattr(e, "loop/autoclear", "0", true);
+        if (r < 0)
+                return r;
```

Comment:
hmm, can you elaborate why this is necessary? Note that the shutdown logic is to continue detaching stuff until nothing can be changed anymore, and accepting EBUSY while looping. Hence, it shouldn't matter if attempt to detach loop devices with autoclear or not, it will just fail with EBUSY if it can't be detached, but that should not have effect on anything as we will stop the loop anyway as soon as everything that remains consistently gives us EBUSY.

loop device autoclear is quite racy btw: when you turn it on it has no immediate effect, only as soon as the next client opens/closes the device. Which is why I think we should continue try to detach it, just in case the device was left in this state...

Reply:
However, loop device autoclear is on by default, i.e. util-linux mount & systemd-mount setup loopmounts with autoclear on and do so correctly.

I happen to mount squashfs with kernel modules, and bind-mounting that into /usr/lib/modules, which systemd-udevd memory-maps and helds up open (we don't stop systemd-udevd on shutdown it seems, LazyUnmount helps but I'd rather not). I have other circular mounts like that as well, since my sysroot (/) also is from a squashfs loop mount from a partition, which is also mounted under / as well.

Meaning that, yes there is consistent EBUSY and I am watching never ending QEMU VM doing "trying to detach loop0, couldn't detach loop0, one mount is still left". And shutdown is not pivoting to the shutdown initrd (which I am providing).

At the very least, I'd like to skip autoclear devices during pre-pivot shutdown (the one off sysroot /) as things are still fairly normal at that point. And for example then we can try clearing them post-pivot.

Let me try to get the screengrabs of the video recordings I did. Or maybe I can try booting this with serial console capture.

Reply:
I guess my point is, if one is continiously failing to unmount some filesystem, and even failed to remount it RO, no amount of detaching loopback devices will help.

I also see things repartition the device which is ultimately under sysroot squashfs.

Reply:
so, again the algorithm is supposed to exit the loop automatically when everything in place is EBUSY and cannot be changed. If that's not working for you the algorithm is borked as a whole and that should be fixed. I'd really like to understand what's going on here.

Normally udevd would be killed as one of the first things in that loop. Why isn't it for you? can you provide a debug log of this shutdown loop (or screenshot?)?

again, if for some reason we keep looping even though we get EBUSY from the loopback device forever then that's something to fix by checking why we continue looping. i.e. fix the borked algorithm, not the reason we entered the borked algorithm.

And yes, if we continously fail at unmounting/detaching/killing and nothing changes anymore we *do* leave the loop already and give up. You will see a message such as ""Cannot finalize remaining file systems, …, continuing." and then we leave the loop:

https://github.com/systemd/systemd/blob/master/src/shutdown/shutdown.c#L467

Reply:
With this change in place, things happen like that "Not all file system sunmounted, 1 lef. Cannot finalize remaining file systems, continuing. Successfully change into root pivot."

Now let me get back the shutdown loop without this change in. Rebuilding the VM.

Reply:
![shutdown-loop](https://user-images.githubusercontent.com/19779/71178174-86648f80-2265-11ea-9d80-a09330a65f23.png)

Reply:
Since I see Detaching loopback /dev/loop0 and Not all loop devices detached, it means delete_loopback returned 1.

Meaning r = ioctl(fd, LOOP_CLR_FD, 0); returned >= 0.

Reply:
Omg, i don't like the comments in loop_clr_fd in the kernel.

Reply:
Hmmm i think

```
                if (k >= 0 &&
                    major(root_st.st_dev) != 0 &&
                    lstat(m->path, &loopback_st) >= 0 &&
                    root_st.st_dev == loopback_st.st_rdev) {
                        n_failed++;
                        continue;
                }
```

this is also can cause this.

Reply:
rebuilding with more debug statements.

Reply:
hmm, any chance you can get the output above, but with debug logging? should suffice hitting "systemd-analyze log-level debug" before "poweroff"

Reply:
so i stared at this for a while now, couldn't figure anything out. But I prepped #14409 now, which fixes some minor things and modernizes stuff, but shouldn't effect your bug.

I am puzzled by your bug. Would love to see the loop output when debug logging is on, otherwise not sure what I can help...

Reply:
arrrgggh

we are working on this in parallel.

I have better code issue explanation for this, without transversing and guessing things.

Let me finish typing up the reproducer & issue at hand.

Reply:
@poettering https://github.com/systemd/systemd/issues/14410